exporter/loadbalancingexporter/resolver_dns.go

```diff
@@ -0,0 +1,172 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package loadbalancingexporter
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sort"
+	"sync"
+	"time"
+
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+	"go.uber.org/zap"
+)
+
+var _ resolver = (*dnsResolver)(nil)
+
+const (
+	defaultResInterval = 5 * time.Second
+	defaultResTimeout  = time.Second
+)
+
+var errNoHostname = errors.New("no hostname specified to resolve the backends")
+
+type dnsResolver struct {
+	logger *zap.Logger
+
+	hostname    string
+	resolver    netResolver
+	resInterval time.Duration
+	resTimeout  time.Duration
+
+	endpoints         []string
+	onChangeCallbacks []func([]string)
+
+	stopCh     chan (struct{})
+	updateLock sync.Mutex
+	shutdownWg sync.WaitGroup
+}
+
+type netResolver interface {
+	LookupIPAddr(ctx context.Context, host string) ([]net.IPAddr, error)
+}
+
+func newDNSResolver(logger *zap.Logger, hostname string) (*dnsResolver, error) {
+	if len(hostname) == 0 {
+		return nil, errNoHostname
+	}
+
+	return &dnsResolver{
+		logger:      logger,
+		hostname:    hostname,
+		resolver:    &net.Resolver{},
+		resInterval: defaultResInterval,
+		resTimeout:  defaultResTimeout,
+		stopCh:      make(chan struct{}),
+	}, nil
+}
+
+func (r *dnsResolver) start(ctx context.Context) error {
+	if _, err := r.resolve(ctx); err != nil {
+		return err
+	}
+
+	go r.periodicallyResolve()
+
+	return nil
+}
+
+func (r *dnsResolver) shutdown(ctx context.Context) error {
+	r.onChangeCallbacks = nil
```

Comment:
I think we need to protect this field, this is a data race with line 150 which read it.

Reply:
Done